### ISSUE TITLE
Missing d3d12video.h?

### ISSUE THREAD

**Issue by ghost** [mingw-w64/mingw-w64#24]

**Comment by Biswa96** [comment]
I can try to work on that. Little questions. How do I know that feature in ffmpeg is compiled and enabled successfully? Would that feature be added in the official ffmpeg repository?

**Comment by Biswa96** [comment]
So, what did happen? Did that feature get merged?

**Comment by oscarbg** [comment]
May be because of new av1 encode support present in d3d12video,h but only in agility sdk..

**Comment by Biswa96** [comment]
Which ffmpeg source code repository should I test for adding d3d12video.h? The patches in intel/cartwheel-ffmpeg repository does not apply.

**Comment by tong1wu** [comment]
> Which ffmpeg source code repository should I test for adding d3d12video.h? The patches in intel/cartwheel-ffmpeg repository does not apply. @tong1wu

Could you please try it again? This happens once in a while but we maintain the patches appliable daily.

**Comment by Biswa96** [comment]
Thanks. All patches can be applied now.

This is the first patch to add d3d12video.idl in wine https://gitlab.winehq.org/wine/wine/-/merge_requests/3927 and I shall let you inform when all things are in upstream.

**Comment by Biswa96** [comment]
All the required d3d12video symbols are in wine repository. It will be imported to mingw-w64 soon™️ 

I change some lines in ffmpeg code to mitigate compiler errors and warnings. Here is the [ffmpeg-test.diff.txt] diff based on your ffmpeg repository d3d12va_encode branch. Please verify the changes first.

[ffmpeg-test.diff.txt]: https://github.com/mingw-w64/mingw-w64/files/12745152/ffmpeg-test.diff.txt

**Comment by tong1wu** [comment]
> All the required d3d12video symbols are in wine repository. It will be imported to mingw-w64 soon™️
> 
> I change some lines in ffmpeg code to mitigate compiler errors and warnings. Here is the [ffmpeg-test.diff.txt](https://github.com/mingw-w64/mingw-w64/files/12745152/ffmpeg-test.diff.txt) diff based on your ffmpeg repository d3d12va_encode branch. Please verify the changes first.

Thanks. It looks good to me. I'll update in next version for upstream.

**Comment by tong1wu** [comment]
> All the required d3d12video symbols are in wine repository. It will be imported to mingw-w64 soon™️
> 
> I change some lines in ffmpeg code to mitigate compiler errors and warnings. Here is the [ffmpeg-test.diff.txt](https://github.com/mingw-w64/mingw-w64/files/12745152/ffmpeg-test.diff.txt) diff based on your ffmpeg repository d3d12va_encode branch. Please verify the changes first.

-    ID3D12Device_GetCustomHeapProperties(ctx->hwctx->device, &heap_props, 0, heap_type);
+    ctx->hwctx->device->lpVtbl->GetCustomHeapProperties(ctx->hwctx->device, &heap_props, 0, heap_type);

May I ask why we need this change? Thx.

**Comment by Biswa96** [comment]
It was an workaround. Sadly, I can not explain it properly. But I can ping someone to explain it.

tl;dr There is an ABI mismatch between MSVC and gcc when the virtual member function returns a structure. So, wine/widl workarounds it with some tricks. See the `WIDL_EXPLICIT_AGGREGATE_RETURNS` magic in d3d12video.h header file in mingw-w64.

**Comment by tong1wu** [comment]
> It was an workaround. Sadly, I can not explain it properly. But I can ping someone to explain it.
> 
> tl;dr There is an ABI mismatch between MSVC and gcc when the virtual member function returns a structure. So, wine/widl workarounds it with some tricks. See the `WIDL_EXPLICIT_AGGREGATE_RETURNS` magic in d3d12video.h header file in mingw-w64.

FFmpeg is a C-style project so I would prefer keeping the original one. Is there any way to have this workaround elsewhere?

**Comment by Biswa96** [comment]
I have compiled the cartwheel-ffmpeg now and that change is not required. I forgot why I changed that code but I remember seeing the same changes in sdl2 and mesa. My apology for any confusion 🙇

**Comment by tong1wu** [comment]
> I have compiled the cartwheel-ffmpeg now and that change is not required. I forgot why I changed that code but I remember seeing the same changes in sdl2 and mesa. My apology for any confusion 🙇

No worries. Glad to hear that.

**Comment by tong1wu** [comment]
Hi there. Thanks for the efforts. Now the d3d12va decoder can be built in MinGW. However, when I'm developing the d3d12va hevc encoder it seems the new added d3d12video.h still lacks of some structures which should have been added too for example, D3D12_FEATURE_DATA_VIDEO_ENCODER_RESOURCE_REQUIREMENTS and D3D12_FEATURE_DATA_VIDEO_ENCODER_RESOLUTION_SUPPORT_LIMITS. I believe D3D12_FEATURE_VIDEO_ENCODER_CODEC and these structures were added to DirectX header at the same time. So we better have these added in our d3d12video.h. Please refer to this https://github.com/intel-media-ci/ffmpeg/pull/689 for more info. Thanks in advance.

**Comment by Biswa96** [comment]
Certainly, I will make an attempt to include the necessary symbols in the header. However, I would like to inform that wine (and vkd3d possibly) is currently in a feature freeze state. It is uncertain whether the merge request(s) will require a longer duration.

**Comment by Biswa96** [comment]
I have sent a merge request to add those symbols https://gitlab.winehq.org/wine/wine/-/merge_requests/4782

**Comment by Biswa96** [comment]
New symbols are available in this ed05999bf5be7847427800d09e847febdde00222 commit.

**Comment by Biswa96** [comment]
This issue can be closed now.

**Comment by Biswa96** [comment]
> As far as I know, d3d12video is still outdated, currently lack the hevc rext GUID.

Could you provide the steps to reproduce the issue?

**Comment by Biswa96** [comment]
It is possible to add those but upstream projects (wine and mingw-w64) may not accept random symbols without testing. So, I asked for a valid widely used project which requires those.

**Comment by Biswa96** [comment]
Could you provide the steps to build it?

**Comment by Biswa96** [comment]
> Or a ffmpeg [fork](https://github.com/cgutman/FFmpeg/tree/dxva_hevc_rext)

I have compiled that ffmpeg fork using msys2 ucrt64 toolchain without any issue. The ffmpeg build options are mentioned here https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-ffmpeg/PKGBUILD + `--enable-d3d12va`

**Comment by Biswa96** [comment]
> This is not the case for [d3d12va](https://github.com/FFmpeg/FFmpeg/blob/9bcb86b0fa58f1a5a2a8613065349e26abadb329/libavcodec/d3d12va_hevc.c#L163-L191), where support cannot be added if d3d12video.h does not have these GUIDs.

D3D12_VIDEO_DECODE_PROFILE_HEVC_MAIN and D3D12_VIDEO_DECODE_PROFILE_HEVC_MAIN10 is already present in d3d12video.idl.

**Comment by Biswa96** [comment]
> If in the future ffmpeg add new profiles here for d3d12va, MinGW will need to provide these new GUIDs.

OK, let's wait for ffmpeg to add new profiles. I can not add symbols without any projects or upstream may not accept it (speaking from previous experience).

**Comment by Biswa96** [comment]
> . ffmpeg's d3d11va use [bundled](https://github.com/cgutman/FFmpeg/commit/9531ce6d655ff874634cc51e326cc85937b1f3f6#diff-7a1cc5fd46fae00f41bbe61e2bbf192b1300dceabd9afb6d7c3d28876b9f1c87R46-R53) GUIDs for historical reasons and thus does not require a MinGW header.

I have sent patch to upstream to add those GUIDs and structure https://gitlab.winehq.org/wine/wine/-/merge_requests/6466. Those will be imported to mingw-w64 after being merged.
